# [LAB-03] 1. 프로젝트 개요 - 연습문제

## 준비작업

### 라이브러리 참조

In [1]:
from jussam import load_data
from helpers import *
from pandas import DataFrame
import numpy as np

📦 연세대학교 주영아 교수가 제작한 라이브러리를 사용중입니다.
📧 Email(1): j.purplerose@yonsei.ac.kr
📧 Email(2): j.purplerose@gmail.com
📝 Website: https://juyounga.kr/
🔖 Version: 0.5.19


## 📚 캘리포니아에서 끝난 줄 알았던 품질 점검

### 당신은 지난 단원에서 캘리포니아 주택가격 보고서의 Ⅰ장을 마친 분석가입니다.

보고서를 덮으며 팀장이 말합니다. "1990년 캘리포니아 한 건으로 우리 점검 절차가 검증됐다고는 말할 수 없다."

그래서 받은 자료가 2014년 5월부터 2015년 5월까지 시애틀이 속한 킹 카운티에서 실제로 거래된 주택 기록입니다. 캘리포니아가 구역 단위로 집계된 자료였던 것과 달리, 이 자료는 **주택 한 채가 한 행**이고 침실 수·욕실 수·면적·건축 연도처럼 집 한 채의 특성이 그대로 들어 있습니다.

행을 지우거나 값을 고치기 전에, **원본 그대로** 구조·형식·품질을 점검하고 기술통계량까지 확인하세요.

다만 점검에 들어가기 전에 변수의 성격부터 정리해야 합니다. **워터프론트 여부(`waterfront`)는 0과 1 두 값뿐인 이진형이므로 범주형으로 바꿔 놓고 시작하세요.** 숫자로 두면 평균·중앙값·왜도가 계산되어 연속형 변수들 사이에 섞여 들어가는데, 0과 1에 대한 평균은 비율일 뿐이라 중심이나 꼬리를 논할 대상이 아닙니다. 

우편번호(`zipcode`), 위도, 경도도 마찬가지입니다. 98001부터 98199까지 숫자로 저장되어 있지만 크기가 아니라 **지역을 가리키는 코드**라서, 평균을 내는 순간 의미가 사라집니다. 반면 조망 점수(`view`)·주택 상태 점수(`condition`)·건축 등급 점수(`grade`)처럼 **여러 단계로 나뉜 점수는 연속형 그대로 두고** 점검합니다.

> "캘리포니아에서 쓰던 판정 기준을 그대로 들이대면 이 데이터에서는 무엇이 걸리는지 보고하세요."

### 💻 코드 작성

#### 데이터 가져오기

In [2]:
origin = load_data("kc_house")

# 이진형 변수는 범주형으로 변환 (0/1을 숫자로 두면 연속형 통계에 섞여 들어갑니다)
origin["waterfront"] = origin["waterfront"].astype("category")

# 우편번호는 숫자로 저장되어 있지만 지역을 가리키는 코드이므로 범주형으로 변환
origin["zipcode"] = origin["zipcode"].astype("category")

origin.head()

📚 이 데이터 세트는 시애틀이 속한 킹 카운티의 주택 매매 가격 정보를 담고 있습니다. 2014년 5월부터 2015년 5월까지 판매된 주택들이 포함되어 있습니다.

(출처: https://www.kaggle.com/datasets/harlfoxem/housesalesprediction)

    field          description
--  -------------  ------------------------------------------------
 0  date           거래 날짜
 1  price          주택 가격 (USD 달러)
 2  bedrooms       침실 수 (개수)
 3  bathrooms      욕실 수 (개수)
 4  sqft_living    거주 공간 면적 (제곱 피트)
 5  sqft_lot       대지 면적 (제곱 피트)
 6  floors         층 수 (개수)
 7  waterfront     워터프론트 여부 (이진형: 1/0)
 8  view           조망 점수 (0~4)
 9  condition      주택 상태 점수 (1~5)
10  grade          건축 등급 점수 (1~13)
11  sqft_above     지상 면적 (제곱 피트)
12  sqft_basement  지하 면적 (제곱 피트)
13  yr_built       건축 연도
14  yr_renovated   리모델링 연도
15  zipcode        우편번호
16  lat            위도
17  long           경도
18  sqft_living15  15개 인근 주택의 평균 거주 공간 면적 (제곱 피트)
19  sqft_lot15     15개 인근 주택의 평균 대지 면적 (제곱 피트)



,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,2014-10-13,221900.000,3,1.000,1180,5650,1.000,0,0,3,7,1180,0,1955,0,98178,47.511,-122.257,1340,5650
1,2014-12-09,538000.000,3,2.250,2570,7242,2.000,0,0,3,7,2170,400,1951,1991,98125,47.721,-122.319,1690,7639
2,2015-02-25,180000.000,2,1.000,770,10000,1.000,0,0,3,6,770,0,1933,0,98028,47.738,-122.233,2720,8062
3,2014-12-09,604000.000,4,3.000,1960,5000,1.000,0,0,5,7,1050,910,1965,0,98136,47.521,-122.393,1360,5000
4,2015-02-18,510000.000,3,2.000,1680,8080,1.000,0,0,3,8,1680,0,1987,0,98074,47.617,-122.045,1800,7503


#### 데이터 구조 점검

In [3]:
print(f"데이터 크기: {origin.shape}")
origin.info()

데이터 크기: (21613, 20)
<class 'pandas.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   date           21613 non-null  datetime64[ns]
 1   price          21613 non-null  float64       
 2   bedrooms       21613 non-null  int64         
 3   bathrooms      21613 non-null  float64       
 4   sqft_living    21613 non-null  int64         
 5   sqft_lot       21613 non-null  int64         
 6   floors         21613 non-null  float64       
 7   waterfront     21613 non-null  category      
 8   view           21613 non-null  int64         
 9   condition      21613 non-null  int64         
 10  grade          21613 non-null  int64         
 11  sqft_above     21613 non-null  int64         
 12  sqft_basement  21613 non-null  int64         
 13  yr_built       21613 non-null  int64         
 14  yr_renovated   21613 non-null  int64         
 15  zipcode   

#### 형식·표기·단위 점검 - 연속형 변수

In [4]:
# 값을 검사할 필드 목록 (숫자형 변수 전체)
# --> 범주형으로 바꾼 waterfront·zipcode와 날짜형인 date는 자동으로 빠집니다.
fields = my_qtcheck.get_number_column_names(origin)

# 위도·경도는 주택의 위치를 나타내는 좌표이므로 값의 범위를 점검할 대상이 아닙니다.
fields = [f for f in fields if f not in ['lat', 'long']]
print(f"점검 대상 변수 {len(fields)}개: {fields}")

minmax = []  # 최소값과 최대값을 저장할 리스트

for field in fields:
    min_value = origin[field].min()
    max_value = origin[field].max()
    minmax.append({"min": min_value, "max": max_value})

# 최소값과 최대값을 데이터 프레임으로 변환하여 출력
minmax_df = DataFrame(minmax, index=fields)
minmax_df

점검 대상 변수 15개: ['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'sqft_living15', 'sqft_lot15']


,min,max
price,75000.000,7700000.000
bedrooms,0.000,33.000
bathrooms,0.000,8.000
sqft_living,290.000,13540.000
sqft_lot,520.000,1651359.000
floors,1.000,3.500
view,0.000,4.000
condition,1.000,5.000
grade,1.000,13.000
sqft_above,290.000,9410.000


#### 형식·표기·단위 점검 - 범주형 변수

In [5]:
# 워터프론트 여부 - 두 값의 분포 확인
display(origin['waterfront'].value_counts())

# 우편번호 - 지역 수와 가장 거래가 많은 지역 확인
print(f"우편번호(지역) 수: {origin['zipcode'].nunique()}개")
display(origin['zipcode'].value_counts().head())

waterfront
0    21450
1      163
Name: count, dtype: int64

우편번호(지역) 수: 70개


zipcode
98103    602
98038    590
98115    583
98052    574
98117    553
Name: count, dtype: int64

#### 데이터 품질 점검 - 중복값과 결측치

In [6]:
df1 = my_qtcheck.check_duplicates(origin)

missing = my_qtcheck.check_missing_values(df1)
print(f"결측치가 있는 변수의 수: {(missing['Missing Count'] > 0).sum()}")

# 0으로 기록된 값의 비율 점검
for field in ['yr_renovated', 'sqft_basement', 'bedrooms']:
    zero_ratio = (df1[field] == 0).mean() * 100
    print(f"{field} 가 0인 비율: {zero_ratio:.1f}% ({(df1[field] == 0).sum()}건)")

중복된 행의 수: 0
결측치가 있는 변수의 수: 0
yr_renovated 가 0인 비율: 95.8% (20699건)
sqft_basement 가 0인 비율: 60.7% (13126건)
bedrooms 가 0인 비율: 0.1% (13건)


#### 기술 통계량

In [7]:
desc_df = my_qtcheck.numerical_summary(df1)
desc_df[['mean', '50%', 'rel_diff', 'rdiff_flag', 'skew', 'outliers_ratio', 'log_need']]

,mean,50%,rel_diff,rdiff_flag,skew,outliers_ratio,log_need
price,540088.142,450000.000,0.200,diff,4.024,0.053,log
bedrooms,3.371,3.000,0.124,diff,1.974,0.025,log1p
bathrooms,2.115,2.250,0.060,similar,0.511,0.026,log1p
sqft_living,2079.900,1910.000,0.089,similar,1.472,0.026,log
sqft_lot,15106.968,7618.000,0.983,large_diff,13.060,0.112,log
floors,1.494,1.500,0.004,similar,0.616,0.000,none
view,0.234,0.000,inf,large_diff,3.396,0.098,log1p
condition,3.409,3.000,0.136,diff,1.033,0.001,log
grade,7.657,7.000,0.094,similar,0.771,0.088,log
sqft_above,1788.391,1560.000,0.146,diff,1.447,0.028,log


#### 판정 결과 집계

In [8]:
# 판정 등급별 변수 개수
print("[ rdiff_flag 판정별 변수 개수 ]")
print(desc_df['rdiff_flag'].value_counts())

# 상대 거리가 유한한 숫자로 계산되지 않은 변수
inf_fields = desc_df[np.isinf(desc_df['rel_diff'])]
print(f"\n[ rel_diff 가 무한대인 변수: {len(inf_fields)}개 ]")
print(inf_fields[['mean', '50%', 'rel_diff', 'rdiff_flag']])

# 왜도 상위 5개
print("\n[ 왜도 상위 5개 ]")
print(desc_df['skew'].sort_values(ascending=False).head(5))

[ rdiff_flag 판정별 변수 개수 ]
rdiff_flag
similar       8
large_diff    5
diff          4
Name: count, dtype: int64

[ rel_diff 가 무한대인 변수: 3개 ]
                 mean   50%  rel_diff  rdiff_flag
view            0.234 0.000       inf  large_diff
sqft_basement 291.509 0.000       inf  large_diff
yr_renovated   84.402 0.000       inf  large_diff

[ 왜도 상위 5개 ]
sqft_lot       13.060
sqft_lot15      9.507
yr_renovated    4.549
price           4.024
view            3.396
Name: skew, dtype: float64


### 문제 풀이

#### 1. 값의 범위를 점검하던 중, 주택 한 채의 침실 수로는 도저히 성립하지 않는 기록이 눈에 띄었습니다. 이 데이터에 기록된 침실 수의 최댓값은 얼마인가요? (정수)

- **정답**: `33`
- **복습 개념**: 형식·표기·단위 점검입니다. 값이 물리적으로 가능한 범위 안에 있는지를 최솟값·최댓값으로 훑어보는 단계이며, 캘리포니아에서 상한 절단의 단서를 처음 발견했던 바로 그 단계입니다.
- **풀이 접근**: 숫자형 변수 **전체**를 목록으로 묶어 각각의 최솟값과 최댓값을 구한 뒤 한 표로 모아 봅니다. 그리고 그 표를 변수의 의미와 하나씩 대조합니다. "이 값이 현실에서 가능한가"를 묻는 것이 이 단계의 전부이므로, 검사 목록에서 변수 하나가 빠지면 그 변수의 오류는 끝까지 발견되지 않습니다. 목록을 손으로 적지 않고 숫자형 컬럼 이름을 통째로 가져오는 이유가 여기 있습니다.
- **근거(계산 결과)**: 침실 수의 범위는 **0 ~ 33**입니다. 최댓값 33이 기록된 주택을 열어 보면 거주 면적이 1,620제곱피트(약 45평), 욕실은 1.75개입니다. 45평에 침실 33개는 성립하지 않으므로 **3을 33으로 잘못 입력한 기록**으로 보는 것이 타당합니다.
- **함께 생각해 볼 점**: 같은 표에서 침실 수와 욕실 수의 최솟값이 둘 다 0이라는 점도 눈에 띕니다. 주거용 주택에 침실도 욕실도 없다는 뜻이 되니 이것 역시 정상적인 값은 아닙니다. 다만 이 단계에서는 **발견하고 기록만** 합니다. 고치거나 지우는 것은 뒤의 전처리 단계 일입니다.
- **주의**: 숫자형이라고 해서 전부 점검 대상인 것은 아닙니다. 우편번호(zipcode)는 98001~98199 사이의 숫자로 저장되어 있지만 크기가 아니라 **지역을 가리키는 코드**라서 범주형으로 바꿔 두었고, 위도(lat)·경도(long)는 주택이 어디에 있는지를 가리키는 **좌표**라서 "이 값이 현실에서 가능한가"라는 질문 자체가 성립하지 않아 목록에서 뺐습니다. 그래서 이 표에 오르는 것은 15개입니다. 숫자로 저장되어 있다는 것과 크기를 비교할 수 있다는 것은 다른 이야기입니다.

#### 2. 결측치를 점검했더니 한 건도 없었습니다. 그런데 리모델링 연도가 0으로 기록된 주택이 있습니다. 이런 주택은 전체의 몇 %인가요? (소수 첫째 자리, % 단위)

- **정답**: `95.8`
- **복습 개념**: 0으로 기록된 값의 해석입니다. 결측치 점검표는 비어 있는 칸만 세기 때문에, **"없음"을 0으로 적어 둔 값은 결측으로 잡히지 않습니다.**
- **풀이 접근**: 먼저 결측치 점검표로 비어 있는 칸이 정말 없는지 확인합니다. 그다음 0이 의미를 갖는 변수를 골라 0인 행의 비율을 직접 계산합니다.
- **근거(계산 결과)**: 리모델링 연도가 0인 주택은 20,699건으로 전체 21,613건의 **95.8%** 입니다. 이 0은 "연도를 모른다"가 아니라 **"리모델링한 적이 없다"** 는 뜻입니다. 같은 방식으로 지하 면적이 0인 주택도 60.7%인데, 이것도 결측이 아니라 지하실이 없다는 뜻입니다.
- **자주 하는 실수**: 0을 결측으로 보고 평균으로 대체하면, 리모델링한 적 없는 집에 "1971년쯤 리모델링했다"는 없던 사실이 생겨납니다. 반대로 0을 그대로 두고 평균을 내면 리모델링 연도의 평균이 84년이라는 이상한 숫자가 나옵니다. 이 변수는 연도가 아니라 **리모델링 여부(0/1)** 로 바꿔 쓰는 편이 맞습니다.

#### 3. 평균과 중앙값의 상대적 거리로 자동 판정된 등급 중, 캘리포니아 데이터에서는 해당 변수가 하나도 없었던 가장 심한 등급이 이 데이터에서는 여러 변수에 붙었습니다. 그 등급의 이름은 무엇인가요? (영문 판정값 1개)

- **정답**: `large_diff`
- **복습 개념**: 평균과 중앙값의 상대적 거리에 따른 판정입니다. 거리가 0.1 미만이면 `similar`, 0.1 이상 0.5 미만이면 `diff`, 0.5 이상이면 `large_diff`로 판정됩니다.
- **풀이 접근**: 기술통계량 결과표에서 판정 열만 뽑아 값별로 개수를 세어 봅니다. 캘리포니아에서는 이 열에 `similar`와 `diff`만 있었다는 점을 떠올리며 비교합니다.
- **근거(계산 결과)**: 판정 결과는 `similar` 8개, `large_diff` 5개, `diff` 4개입니다. 캘리포니아에서는 `large_diff`가 **0개**여서 "평균이 완전히 무너진 변수는 없다"고 결론지었는데, 이 데이터에서는 **5개**나 나왔습니다. 대지 면적(sqft_lot)의 상대 거리는 **0.983** 으로, 평균이 중앙값의 거의 두 배입니다. 워터프론트 여부와 우편번호는 범주형으로 바꿔 두었으므로 이 표에 아예 올라오지 않습니다.
- **헷갈리기 쉬운 점**: `large_diff`가 "데이터가 잘못됐다"는 뜻은 아닙니다. 캘리포니아는 구역 단위로 집계된 자료라 값이 어느 정도 평탄해졌지만, 이 자료는 주택 한 채가 한 행이라 소수의 대저택이 평균을 그대로 끌어올립니다. **집계 단위가 다르면 같은 지표도 다르게 나옵니다.**

#### 4. 그 가장 심한 등급을 받은 변수들을 자세히 보면, 일부는 상대적 거리 값 자체가 유한한 숫자로 계산되지 않습니다. 그런 변수는 모두 몇 개인가요? (정수)

- **정답**: `3`
- **복습 개념**: 지표가 성립하지 않는 경우를 알아보는 문제입니다. 상대적 거리는 평균과 중앙값의 차이를 **중앙값으로 나눠** 구하므로, 중앙값이 0이면 0으로 나누게 되어 값이 무한대가 됩니다.
- **풀이 접근**: 기술통계량 결과표에서 상대 거리 열이 유한한 숫자인지 확인하고, 무한대인 행만 걸러 세어 봅니다. 그 행들의 중앙값 열을 함께 보면 원인이 바로 보입니다.
- **근거(계산 결과)**: 무한대로 계산된 변수는 **3개** 입니다 — 조망 점수(view), 지하 면적(sqft_basement), 리모델링 연도(yr_renovated). 세 변수 모두 **중앙값이 0** 입니다. 절반 넘는 주택이 0이라서 중앙값이 0이 되고, 그 결과 나눗셈이 성립하지 않습니다.
- **실무 포인트**: 판정 열에는 `large_diff`라고 찍혀 있지만 이 세 변수는 **판정 자체가 의미 없는 경우**입니다. 결과표의 판정값을 그대로 옮겨 적기 전에 그 값이 어떻게 계산됐는지 확인해야 하는 이유가 여기 있습니다. 세 변수의 0은 크기가 0이라는 뜻이 아니라 조망·지하실·리모델링이 **없다**는 뜻이라, 애초에 평균·중앙값으로 중심을 논할 대상이 아닙니다.
- **함께 생각해 볼 점**: 워터프론트 여부도 중앙값이 0이라 숫자로 두었다면 여기 네 번째로 끼어들었을 변수입니다. 시작할 때 범주형으로 바꿔 두었기 때문에 이 표에서 빠졌고, 그만큼 **판정이 의미 없는 변수를 걸러내는 일을 통계 이전 단계에서 끝낸 셈**입니다.

#### 5. 팀장에게 "이 변수부터 로그 변환을 검토해야 한다"고 보고하려 합니다. 우측 꼬리가 가장 극심한 변수는 무엇인가요? (변수명 1개)

- **정답**: `sqft_lot`
- **복습 개념**: 왜도(skew)로 분포의 꼬리 방향과 정도를 판정하는 단계입니다. 왜도가 0.5보다 크면 우측 꼬리로 판정하며, 값이 클수록 꼬리가 깁니다.
- **풀이 접근**: 기술통계량 결과표의 왜도 열을 내림차순으로 정렬해 가장 큰 값을 찾습니다. 상대적 거리 판정과 로그 변환 필요 여부 열도 함께 보면 근거가 겹쳐 쌓입니다.
- **근거(계산 결과)**: 대지 면적(**sqft_lot**)의 왜도가 **13.060** 으로 1위이고, 2위는 인근 대지 면적(sqft_lot15) **9.507** 입니다. sqft_lot은 상대 거리 0.983으로 `large_diff`이고 이상치 비율도 11.2%로 가장 높아, 세 지표가 모두 같은 곳을 가리킵니다. 참고로 워터프론트 여부를 숫자로 두었다면 왜도 11.385로 2위에 올라왔을 텐데, 0과 1뿐인 값의 왜도는 꼬리가 아니라 **한쪽 값이 드물다는 뜻**일 뿐입니다. 범주형으로 바꿔 두면 이런 가짜 순위가 애초에 생기지 않습니다.
- **결론**: 팀장에게 드릴 답은 이렇게 정리됩니다. **캘리포니아에서는 상한 절단이 문제였지만 이 데이터에는 절단이 없고, 대신 주택 한 채 단위라서 꼬리가 훨씬 깁니다.** 그래서 걸리는 것은 ① 침실 33개 같은 입력 오류, ② 0을 결측으로 오해할 위험, ③ 면적 계열의 극단적인 우측 꼬리 세 가지입니다. 판정 기준은 그대로 쓸 수 있었지만, **기준이 가리키는 문제는 데이터마다 달랐습니다.**